src/formatting/imports.rs

```diff
@@ -197,6 +201,7 @@ pub(crate) fn flatten_use_trees(use_trees: Vec<UseTree>) -> Vec<UseTree> {
             }
             tree
         })
+        .unique_by(|ut| format!("{}", ut))
```

Comment:
I would much rather implement `Hash` on `UseTree` so that we can just use `unique()` directly instead of taking the hit on calling `format`.

We're really just dealing with the path so I think this would suffice (along with the default Hash derive on UseSegment):
```rust
impl Hash for UseTree {
    fn hash<H: Hasher>(&self, state: &mut H) {
        self.path.hash(state);
    }
}
```

Reply:
Before I try to implement this, I would like to make sure fully understand the implementation.  My main question is **what is default Hash derive on `UseSegment`**?  How and were it is implemented?

I am asking because my first approach was to implement `Hasher` for both `UseTree` and `UseSegment`.  This is by generating the list of the segments\` text and then hashing the generated text.  I ended up with something very similar to the `UseTree` `format` and realized that using the `format` is about the same from performance point of view.

One more question for my information (assuming that the answer is short). I didn't fully understand how the the hash function works. Taking as an example the hash function you suggested above. First, it doesn't have a return value, so where is the hash value stored and how it is used?  Second, through the iterations I assume that several numeric hash valuses are calculated for the non-List `UseSegments`. If this is the case, it is not clear to me if and how these hash values are "combined" to form the hash of the initial `UseTree`.

Reply:
> My main question is what is default Hash derive on UseSegment? How and were it is implemented?

I was referring to the [Hash trait](https://doc.rust-lang.org/std/hash/trait.Hash.html), and that we should able to leverage the [derive](https://doc.rust-lang.org/rust-by-example/trait/derive.html) attribute for UseSegment instead of having to manually implement the Hash trait ourselves. I'm not sure we'll be able to leverage `derive` on UseTree because that only works when all the fields themselves have a Hash impl, and I know that not all of the AST nodes do (feel free to check/test though!). Once there's a Hash impl for both UseTree and UseSegment then you can use `unique()` directly, instead of formatting to a temporary String.

> I am asking because my first approach was to implement Hasher for both UseTree and UseSegment. This is by generating the list of the segments` text and then hashing the generated text. I ended up with something very similar to the UseTree format and realized that using the format is about the same from performance point of view.

I don't think there will be a need to implement the [Hasher](https://doc.rust-lang.org/std/hash/trait.Hasher.html) trait on any of our custom types to resolve this issue. My assumption here was that the itertools functions we were hoping to leverage utilized some hash-based data structure for the deduplication feature we needed, and that in turn would require that the items (UseTree in our case) implemented the Hash trait (side note, looks like they do leverage a hash set internally :wink:).

This is indeed the case, as highlighted in the the [itertools docs for the unique/unique_by functions](https://docs.rs/itertools/0.10.0/itertools/trait.Itertools.html#method.unique). If you were to try to just use `unique()` on the current code, you'd presumably get errors complaining about unbound constraints, because our UseTree does not have a Hash implementation.

Your initial proposal works because you're allocating an intermediate String representation for each UseTree (which we have to do again later for the actual emitting of the formatted code) and that satisfies the itertools constraints because String does have a Hash impl. We don't want to do this though because it is unnecessary, and we should simply ensure that we have a Hash impl.

> One more question for my information (assuming that the answer is short). I didn't fully understand how the the hash function works

Sorry, but that's a little too off topic for a code review within rustfmt :smile: Hopefully the above info helps somewhat, but otherwise I'd refer you to materials like the Book, the stdlib docs (https://doc.rust-lang.org/std/index.html), or even the rustc source if you really want to drill into it.

Reply:
Thanks for the detailed response! The reference to stdlib also help as I realized I missed the `finish` method that emits the hash calculation result.

>I would much rather implement Hash on UseTree so that we can just use unique() directly instead of taking the hit on calling format.

I have now change `unique_by` to `unique` using your suggested code.